src/main/java/com/aws/greengrass/status/FleetStatusService.java

```diff
@@ -295,24 +296,36 @@ public void schedulePeriodicFleetStatusDataUpdate(boolean isDuringConnectionResu
                 initialDelay, periodicPublishIntervalSec, TimeUnit.SECONDS);
     }
 
+    private AtomicReference<OverallStatus> getOverAllStatus() {
+        Instant now = Instant.now();
+        AtomicReference<OverallStatus> overAllStatus = new AtomicReference<>(OverallStatus.HEALTHY);
+        this.kernel.orderedDependencies().forEach(service -> {
+            serviceFssTracksMap.put(service, now);
+            overAllStatus.set(getOverallStatusBasedOnServiceState(overAllStatus.get(), service));
+        });
+        return overAllStatus;
+    }
+
     @SuppressWarnings("PMD.UnusedFormalParameter")
     private void handleServiceStateChange(GreengrassService greengrassService, State oldState,
                                           State newState) {
+
+        Set<GreengrassService> erroredComponentSet = new HashSet<>();
+
         // Not reporting status of other components in errored component message because a deployment could
         // be in-progress and other component's fleet config arn may not have updated
         if (newState.equals(State.ERRORED)) {
-            Set<GreengrassService> erroredComponentSet = new HashSet<>();
             erroredComponentSet.add(greengrassService);
+            erroredComponentsSet.add(greengrassService);
 
             // Evaluate overall status based on current state of all components
-            Instant now = Instant.now();
-            AtomicReference<OverallStatus> overAllStatus = new AtomicReference<>(OverallStatus.HEALTHY);
-            this.kernel.orderedDependencies().forEach(service -> {
-                serviceFssTracksMap.put(service, now);
-                overAllStatus.set(getOverallStatusBasedOnServiceState(overAllStatus.get(), service));
-            });
+            uploadFleetStatusServiceData(erroredComponentSet, getOverAllStatus().get(), null, Trigger.COMPONENT_STATUS_CHANGE);
+            return;
+        }
 
-            uploadFleetStatusServiceData(erroredComponentSet, overAllStatus.get(), null, Trigger.ERRORED_COMPONENT);
+        // if the reported errored component is covered, remove the component from the set and update the component status
+        if (greengrassService.reachedDesiredState() && erroredComponentsSet.remove(greengrassService) ) {
+            uploadFleetStatusServiceData(erroredComponentSet, getOverAllStatus().get(), null, Trigger.COMPONENT_STATUS_CHANGE);
             return;
```

Comment:
1. If a deployment is in progress, should this message be sent? What happens when an errored component is being reinstalled?
2. If the component is broken, you will also need to remove it from errored component set.
3. Is there any scenario where you want to send this message but reachedDesiredState is false?

Reply:
1. Right it shouldn't. When an errored component is being reinstalled, that means either Nucleus restart or another deployment. If Nucleus restarts, the list will be emptied out and start over. If there is another deployment, the component will be removed already since it will either enter broken or desired state from last deployment. Deployment can't be cancelled when component starts install (after config merge)
2. Yes, will update it
3. The only scenario we consider to use this message is when a RUNNING component gets errored and being recovered. And here we define recovery is when the component finish it's lifecycle and reached RUNNING state again. So I think the whole story needs to rely on reachedDesiredState is true